docs/source/configuration.rst
Configuration
=============

Connections to external services and routing rules are defined using a
configuration files which can be in either in JSON (.json) or in YAML
(.yaml, .yml) format. For concrete configuration examples take a look at the
:doc:`examples` section.

There are three main components which should be configured when rabbit_force
is started, likewise the configuration file has three main sections.

* `source`_ - The Salesforce org(s), or in other words message source(s), that
  rabbit_force needs to connect to in order to receive event messages through
  the Streaming API.
* `sink`_ - RabbitMQ broker(s), or in other words message sink(s), to which
  rabbit_force should forward messages using AMQP.
* `router`_ - Routing rules that define to which broker and exchange the
  messages need to be forwarded.

source
------

The `source`_ mapping has two properties, `orgs`_ where the connected
Salesforce orgs can be defined and and an optional `replay`_ property which can
be defined if you want to take advantage of Salesforce's
`message durability <replay_>`_ feature.

orgs
~~~~

The `orgs`_ property's value should be a mapping Salesforce org names and the
orgs' specifications. You can define multiple `name` and `org specification`
pairs, and use the `name` strings in the `router`_ section to refer to specific
Salesforce orgs.

.. code-block:: yaml
    :caption: Multiple Salesforce orgs example

    orgs:
        # a Salesforce org named as "org1"
        org1:
          consumer_key: "<consumer_key1>"
          consumer_secret: "<consumer_secret1>"
          username: "<username1>"
          password: "<password1>"
          resources:
            - type: PushTopic
              spec:
                Name: lead_changes
                ApiVersion: 42.0
                NotifyForFields: Referenced
                NotifyForOperationCreate: true
                NotifyForOperationUpdate: true
                NotifyForOperationDelete: true
                NotifyForOperationUndelete: true
                Query: SELECT Id, Email, Name, Phone, MobilePhone, Status, LeadSource FROM Lead

        # a Salesforce org named as "org2"
        org2:
          consumer_key: "<consumer_key2>"
          consumer_secret: "<consumer_secret2>"
          username: "<username2>"
          password: "<password2>"
          resources:
            - type: PushTopic
              spec:
                Name: contact_changes
                ApiVersion: 42.0
                NotifyForFields: Referenced
                NotifyForOperationCreate: true
                NotifyForOperationUpdate: true
                NotifyForOperationDelete: true
                NotifyForOperationUndelete: true
                Query: SELECT Id, Email, Name, Phone, MobilePhone FROM Contact

The properties of the Salesforce org specification are documented below.

consumer_key
^^^^^^^^^^^^

The `consumer_key's <consumer_key_>`_ value should be the consumer key from
the Salesforce connected app definition.

consumer_secret
^^^^^^^^^^^^^^^

The `consumer_secret's <consumer_secret_>`_ value should be the consumer secret
from the Salesforce connected app definition.

username
^^^^^^^^

The Salesforce user's username as a string.

password
^^^^^^^^

The Salesforce user's password as a string.

resources
^^^^^^^^^

The `resources`_ property's value should be a list of Salesforce REST API
resources, which are capable of generating event messages. For every resource
the `type`_, `spec`_ and optionally the `durable <durable_section_>`_
properties should be defined.

type
""""

The type of the resource should be either `PushTopic <PushTopic spec_>`_ or
`StreamingChannel <StreamingChannel spec_>`_.

spec
""""

A resource specification is a mapping, which should either contain
the full resource definition, which can be used to create the given resource
on the given Salesforce org, or if it's an existing resource, then a single
property should be given which uniquely defines the resource, such as the
``Id`` or the ``Name`` property.

.. code-block:: yaml
    :caption: Existing resource identified by name

    resources:
        - type: PushTopic
          spec:
            Name: lead_changes

If a new resource is defined and the `type`_ of the resource is ``PushTopic``
then the value of the `spec`_ property should conform to the mapping described
in `PushTopic spec`_, if the `type`_ of the resource is ``StreamingChannel``
then the `spec`_ property should conform to `StreamingChannel spec`_.

PushTopic spec
''''''''''''''

When defining a resource of `type`_ ``PushTopic`` then the `spec`_ mapping
can contain exactly the same value's as defined by `Salesforce's REST API
reference for creating PushTopics <PushTopicRef_>`_.

========================== ============= ======================================
Property                   Property type Description
========================== ============= ======================================
Id                         String        Globally unique string that identifies
                                         a record. The Id
                                         property should be defined
                                         **only when** you're trying to refer
                                         to an existing PushTopic.
Name                       String        Required. Descriptive name of the
                                         PushTopic, such as MyNewCases or
                                         TeamUpdatedContacts. Limit: 25
                                         characters. This value identifies the
                                         channel and must be unique.
ApiVersion                 Float         Required. API version to use for
                                         executing the query specified in Query.
                                         It must be an API version greater than
                                         20.0. If your query applies to a
                                         custom object from a package, this
                                         value must match the package's
                                         ApiVersion.
                                         Example value: 43.0
IsActive                   Boolean       Indicates whether the record currently
                                         counts towards the organization's
                                         allocation.
NotifyForFields            String        Specifies which fields are evaluated
                                         to generate a notification.
                                         Valid values: All, Referenced
                                         (default), Select, Where
Description                String        Description of the PushTopic. Limit:
                                         400 characters
NotifyForOperationCreate   Boolean       true if a create operation should
                                         generate a notification, otherwise,
                                         false. Defaults to true. This field is
                                         available in API version 29.0 and
                                         later.
NotifyForOperationUpdate   Boolean       true if an update operation should
                                         generate a notification, otherwise,
                                         false. Defaults to true. This field is
                                         available in API version 29.0 and
                                         later.
NotifyForOperationDelete   Boolean       true if a delete operation should
                                         generate a notification, otherwise,
                                         false. Defaults to true. This field is
                                         available in API version 29.0 and
                                         later.
NotifyForOperationUndelete Boolean       true if an undelete operation should
                                         generate a notification, otherwise,
                                         false. Defaults to true. This field is
                                         available in API version 29.0 and
                                         later.
NotifyForOperations        String        Specifies which record events may
                                         generate a notification. Valid values:
                                         All (default), Create, Extended,
                                         Update.
                                         In API version 29.0 and later, this
                                         field is read-only, and will not
                                         contain information about delete and
                                         undelete events. Use
                                         NotifyForOperationCreate,
                                         NotifyForOperationDelete,
                                         NotifyForOperationUndelete and
                                         NotifyForOperationUpdate to specify
                                         which record events should generate a
                                         notification.

                                         A value of Extended means that neither
                                         create or update operations are set to
                                         generate events.
Query                      String        Required. The SOQL query statement
                                         that determines which record changes
                                         trigger events to be sent to the
                                         channel. Limit: 1,300 characters
========================== ============= ======================================

StreamingChannel spec
'''''''''''''''''''''

When defining a resource of `type`_ ``StreamingChannel`` then the `spec`_
mapping can contain exactly the same value's as defined by `Salesforce's REST
API reference for creating StreamingChannels <StreamingChannelRef_>`_.

=========== ============= =====================================================
Property    Property type Description
=========== ============= =====================================================
Id          String        Globally unique string that identifies a
                          StreamingChannel record. The Id property should be
                          defined **only when** you're trying to refer to an
                          existing StreamingChannel.
Name        String        Required. Descriptive name of the StreamingChannel.
                          Limit: 80 characters, alphanumeric and “_”, “/”
                          characters only. Must start with “/u/”. This value
                          identifies the channel and must be unique.
Description String        Description of the StreamingChannel. Limit: 255
                          characters.
=========== ============= =====================================================

.. _durable_section:

durable
"""""""

Durable is an optional property, if ``true`` then the resource will be deleted
when the service is terminated, if ``false`` then the resource will remain on
server after the service is terminated. The default value is ``true``.

replay
~~~~~~

Replay is an optional property which can be specified to take advantage of
the Streaming API's `message durability <replay_>`_ feature.

Salesforce stores events for 24 hours. Events outside the 24-hour retention
period are discarded. Salesforce extends the event messages with `replayId`
and `createdDate` fields which are called as `ReplayMarkers`. By storing
these `ReplayMarker` values and sending them when subscribing to channels,
message loss can be avoided which might occur due to hardware, software or
network failures.

rabbit_force can store `ReplayMarker` values in a
`Redis <redis_>`_ database. To define the connection to the database the
`address`_ and optionally the `key_prefix`_ properties should be specified.

address
^^^^^^^

The address of the `Redis <redis_>`_ database as a string.

key_prefix
^^^^^^^^^^

A prefix string to add to all keys stored in the database by rabbit_force.

sink
----

The `sink`_ mapping has a single property named `brokers`_ where the connected
RabbitMQ brokers can be defined.

brokers
~~~~~~~

The `brokers`_ property's value should be a mapping RabbitMQ broker
names and the brokers' specifications. You can define multiple `name` and
`broker specification` pairs, and use the `name` strings in the `router`_
section to refer to specific RabbitMQ brokers.

.. code-block:: yaml
    :caption: Multiple RabbitMQ brokers example

    brokers:
        # a RabbitMQ broker named as "broker1"
        broker1:
          host: host1
          exchanges:
            - exchange_name: my_exchange
              type_name: topic
              durable: true

        # a RabbitMQ broker named as "broker2"
        broker2:
          host: host2
          exchanges:
            - exchange_name: my_exchange
              type_name: topic
              durable: true

The properties of the RabbitMQ broker specification are documented below.

host
^^^^

The host name of the broker as a string.

port
^^^^

*Optional* port name of the broker as an integer. If not specified then the
default port number of RabbitMQ will be used which is either ``5672`` if
`ssl`_ is ``false`` or ``5671`` if `ssl`_ is ``true``.

login
^^^^^

*Optional* username as a string. The default value is ``guest``.

password
^^^^^^^^

*Optional* password as a string. The default value is ``guest``.

virtualhost
^^^^^^^^^^^

*Optional* AMQP virtualhost as a string to use for the connection. The default
value is ``/``.

ssl
^^^

*Optional* boolean value. If ``true`` then the service creates an SSL
connection instead of a plain unencrypted one. The default value is ``false``.

verify_ssl
^^^^^^^^^^

*Optional* boolean value. If ``true`` then the server's SSL certificate
will be verified on connection if the `ssl`_ property's value is ``true``. The
default value is ``true``.

login_method
^^^^^^^^^^^^

*Optional* string which defines the AMQP authentication method. The default
value is ``AMQPLAIN``. All the alternative authentication methods are listed
in `RabbitMQ's Authentication <rabbitmq_auth_>`_ documentation.

insist
^^^^^^

*Optional* boolean value. If ``true`` then insist on connecting to a server.
The default value is ``false``.

exchanges
^^^^^^^^^

The `exchanges`_ property's value should be a list of RabbitMQ exchange
definitions. These are the exchanges that can be used in the `router`_ section
by referring to them by their `exchange_name <exchange_name_section_>`_
property.

.. _exchange_name_section:

exchange_name
"""""""""""""

The name of the exchange as a string.

type_name
"""""""""

The exchange's type as a string. The possible values are ``fanout``,
``direct``, ``topic`` and ``headers``. The routing capabilities of the various
exchange types are explained in
`RabbitMQ's documentation <rabbitmq_exchange_type_>`_.

passive
"""""""

*Optional* boolean value. If set, the server will reply with Declare-Ok if
the exchange already exists with the same name, and raise an error if not. The
default value is ``false``.

durable
"""""""

*Optional* boolean value. If set when creating a new exchange, the exchange
will be marked as durable. Durable exchanges remain active when a server
restarts. The default value is ``false``.

auto_delete
"""""""""""

*Optional* boolean value. If set, the exchange is deleted when all queues have
finished using it. The default value is ``false``.

no_wait
"""""""

*Optional* boolean value. If set, the server will not respond to the method.
The default value is ``false``.

arguments
"""""""""

An *Optional* dictionary which can contain
`AMQP arguments <rabbitmq_policies_>`_ to be passed when
creating the exchange. The default value is ``null``.

router
------

The `router`_ section defines how and where to forward incoming messages.

default_route
~~~~~~~~~~~~~

The `default_route`_ property's value should be either a mapping containing
route parameters or ``null``. The `default_route`_ is used to forward messages
without a matching routing rule from the `rules`_ property. If its value is
``null`` then the messages without a matching rule will be dropped.

broker_name
^^^^^^^^^^^

The name of the RabbitMQ broker as a string. Its value should be one of
broker names defined in `brokers`_.

exchange_name
^^^^^^^^^^^^^

The name of the RabbitMQ exchange as a string. Its value should be one of the
`exchange_name`_ properties defined for the broker referred in `broker_name`_.

routing_key
^^^^^^^^^^^

The message's routing key as a string.

properties
^^^^^^^^^^

An *optional* mapping of message properties. RabbitMQ accepts the
``content_type``, ``content_encoding``, ``headers``, ``delivery_mode``,
``priority``, ``correlation_id``, ``reply_to``, ``expiration``, ``message_id``,
``timestamp``, ``type``, ``user_id``, ``app_id`` and ``cluster_id`` properties.
However ``content_type`` and ``content_encoding`` will be overwritten by
rabbit_force so all messages are forwarded with the ``content_type`` of
``application/json`` and with a ``content_encoding`` of ``utf-8``.

rules
~~~~~

The `rules`_ property is an *optional* list of routing rules. Every rule should
contain a `condition`_ and `route`_ property. On receiving a message,
rabbit_force will iterate over the defined routing `rules`_ and will use the
`route`_ from the first matching rule to forward the message. If no matching
rule is found, then the `default_route`_ will be used.

condition
^^^^^^^^^

The `condition`_ property of the routing rule is used to decide whether the
rule should be applied to a given message or not. The value of this property
should be a `JSONPath <jsonpath_>`_ expression as a string. The JSONPath

CHANGELOG.rst
Changelog
=========

0.1.0 (2018-07-05)
------------------

- Forward `Streaming API <api_>`_ messages from one or more Salesforce orgs to
  one or more `RabbitMQ brokers <rabbitmq_>`_
- Route incoming messages to a specific broker and exchange with the
  specified routing key and properties, with the help of routing rules defined
  as `JSONPath <jsonpath_>`_ expressions
- Support for Salesforce's replay extension for `message reliability and
  durability <replay_>`_ by storing replay markers in a `Redis <redis_>`_
  database
- Configurable error handling behavior, either fail instantly or try to recover
  from network and service outages
- Message sources, sinks and routing configurable with JSON or YAML
  configuration files
- Implemented using `python asyncio <asyncio_>`_ for efficient handling of
  IO intensive operations

.. _aiohttp: https://github.com/aio-libs/aiohttp/
.. _aiocometd: https://github.com/robertmrk/aiocometd/
.. _asyncio: https://docs.python.org/3/library/asyncio.html
.. _api: https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/intro_stream.htm
.. _PushTopic: https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/working_with_pushtopics.htm
.. _GenericStreaming: https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/generic_streaming_intro.htm#generic_streaming_intro
.. _replay: https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/using_streaming_api_durability.htm
.. _CometD: https://cometd.org/
.. _Comet: https://en.wikipedia.org/wiki/Comet_(programming)
.. _Bayeux: https://docs.cometd.org/current/reference/#_bayeux
.. _ext: https://docs.cometd.org/current/reference/#_bayeux_ext
.. _password_auth: https://help.salesforce.com/articleView?id=remoteaccess_oauth_username_password_flow.htm&type=5
.. _refresh_auth: https://help.salesforce.com/articleView?id=remoteaccess_oauth_refresh_token_flow.htm&type=5
.. _connected_app: https://help.salesforce.com/articleView?id=connected_app_overview.htm&type=5
.. _sf_auth: https://help.salesforce.com/articleView?id=remoteaccess_authenticate_overview.htm
.. _web_server_auth: https://help.salesforce.com/articleView?id=remoteaccess_oauth_web_server_flow.htm&type=5
.. _rabbitmq: http://www.rabbitmq.com/
.. _microservice: http://microservices.io/patterns/communication-style/messaging.html
.. _jsonpath: http://goessner.net/articles/JsonPath/
.. _redis: https://redis.io/


rabbit_force/config.py
"""Marshmallow schemas for configuration validation and functions for
loading configurations from files"""
from pathlib import Path
import json
import logging

from marshmallow import Schema, fields, validates_schema, ValidationError, \
    post_load
from marshmallow.validate import Length, Range, OneOf
import yaml

from .salesforce import StreamingResourceType
from .exceptions import ConfigurationError


LOGGER = logging.getLogger(__name__)


class StrictSchema(Schema):
    """Common schema base class which rejects unknown fields"""

    # pylint: disable=unused-argument
    @validates_schema(pass_original=True)
    def check_unknown_fields(self, data, original_data):
        """Check for the presence and reject unknown fields

        :raise marshmallow.exceptions.ValidationError: If an unknown field is \
        found
        """
        # get the difference of the loaded and specified fields
        unknown_fields = set(original_data) - set(self.fields)
        # raise an error if any surplus fields are present
        if unknown_fields:
            raise ValidationError('Unknown field', list(unknown_fields))

    # pylint: enable=unused-argument


class PushTopicSchema(StrictSchema):
    """Configuration schema for PushTopic resources"""

    # PushTopic fields are validated according to
    # https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/\
    # api_streaming/pushtopic.htm
    Id = fields.String(validate=Length(min=1))
    Name = fields.String(validate=Length(min=1, max=25))
    ApiVersion = fields.Float(validate=Range(min=20.0, max=42.0))
    IsActive = fields.Boolean(default=True)
    NotifyForFields = fields.String(validate=OneOf(["All",
                                                    "Referenced",
                                                    "Select",
                                                    "Where"]))
    Description = fields.String(default=None, validate=Length(max=400))
    NotifyForOperationCreate = fields.Boolean(default=True)
    NotifyForOperationUpdate = fields.Boolean(default=True)
    NotifyForOperationDelete = fields.Boolean(default=True)
    NotifyForOperationUndelete = fields.Boolean(default=True)
    NotifyForOperations = fields.String(validate=OneOf(["All",
                                                        "Create",
                                                        "Extended",
                                                        "Update"]))
    Query = fields.String(validate=Length(min=1, max=1300))

    @validates_schema
    def check_required_fileds(self, data):  # pylint: disable=no-self-use
        """Check for required fields

        :raise marshmallow.exceptions.ValidationError: If no fields are \
        specified or if only a single non identifier field is specified or \
        multiple fields are specified but they're not enough for a resource \
        definition
        """
        if len(data) == 1:
            unique_id_fields = {"Id", "Name"}
            if not data.keys() & unique_id_fields:
                raise ValidationError("If only a single field is specified "
                                      "it should be a unique identifier like "
                                      "'Id' or 'Name'.")
        elif len(data) > 1:
            required_fields = {"Name", "ApiVersion", "Query"}
            if (data.keys() & required_fields) != required_fields:
                raise ValidationError("If multiple fields are specified it "
                                      "it should be a full resource "
                                      "definition where at least 'Name', "
                                      "'ApiVersion' and 'Query' are required.")
        else:
            raise ValidationError("Either a single fields should be specified "
                                  "which uniquely identifies the resource or "
                                  "multiple fields which can be used to "
                                  "construct the resource.")

    @validates_schema
    def check_api_version(self, data):  # pylint: disable=no-self-use
        """Check for invalid fields for the specified API version

        :raise marshmallow.exceptions.ValidationError: If any invalid fields \
        found for the specified API version
        """
        # skip validation if the ApiVersion field is not present, which might
        # happen even when it's specified but it's value fails on validation
        if "ApiVersion" not in data:
            return

        # check for the presence of old fields for a newer API version
        if (data["ApiVersion"] >= 29.0 and
                "NotifyForOperations" in data):
            raise ValidationError("'NotifyForOperations' can only be specified"
                                  " for API version 28.0 and earlier.")

        # check for the presence of new fields for an older API version
        elif (data["ApiVersion"] <= 28.0 and
              ("NotifyForOperationCreate" in data or
               "NotifyForOperationDelete" in data or
               "NotifyForOperationUndelete" in data or
               "NotifyForOperationUpdate" in data)):
            raise ValidationError("'NotifyForOperationCreate', "
                                  "'NotifyForOperationDelete', "
                                  "'NotifyForOperationUndelete' and "
                                  "'NotifyForOperationUpdate' can only be "
                                  "specified for API version 29.0 and later.")


class StreamingChannelSchema(StrictSchema):
    """Configuration schema for StreamingChannel resources"""

    # StreamingChannel fields are validated according to
    # https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/\
    # api_streaming/streamingChannel.htm
    Id = fields.String(validate=Length(min=1))
    Name = fields.String(validate=Length(min=1, max=80))
    Description = fields.String(default=None, validate=Length(max=255))

    @validates_schema
    def check_required_fileds(self, data):  # pylint: disable=no-self-use
        """Check for required fields

        :raise marshmallow.exceptions.ValidationError: If no fields are \
        specified or if only a single non identifier field is specified or \
        multiple fields are specified but they're not enough for a resource \
        definition
        """
        if len(data) == 1:
            unique_id_fields = {"Id", "Name"}
            if not data.keys() & unique_id_fields:
                raise ValidationError("If only a single field is specified "
                                      "it should be a unique identifier like "
                                      "'Id' or 'Name'.")
        elif not data:
            raise ValidationError("Either a single fields should be specified "
                                  "which uniquely identifies the resource or "
                                  "multiple fields which can be used to "
                                  "construct the resource.")


class StreamingResourceSchema(StrictSchema):
    """Configuration schema for streaming resources"""

    type = fields.String(required=True, attribute="resource_type",
                         validate=OneOf([_.value for _ in
                                         StreamingResourceType]))
    spec = fields.Dict(required=True, attribute="resource_spec")
    durable = fields.Boolean()

    @post_load
    def load_spec(self, data):
        """Load the spec field with the appropriate schema based on the
        type field"""
        # resource type scpecific schema classes
        schema_map = {
            StreamingResourceType.PUSH_TOPIC: PushTopicSchema,
            StreamingResourceType.STREAMING_CHANNEL: StreamingChannelSchema
        }

        # get the resource type value
        type_name = data[self.fields["type"].attribute]

        # get the schema class for the resource type
        schema_cls = schema_map[type_name]

        # load and update the value of spec field
        spec = data[self.fields["spec"].attribute]
        data[self.fields["spec"].attribute] = schema_cls().load(spec)

        # return the updated data
        return data


class SalesforceOrgSchema(StrictSchema):
    """Configuration schema for a Salesforce organization"""
    consumer_key = fields.String(required=True)
    consumer_secret = fields.String(required=True)
    username = fields.String(required=True)
    password = fields.String(required=True)
    resources = fields.List(fields.Nested(StreamingResourceSchema()),
                            required=True,
                            validate=Length(min=1),
                            attribute="streaming_resource_specs")


class ReplaySchema(StrictSchema):
    """Configuration schema for a Redis replay marker storage"""
    address = fields.Url(schemes=("redis",), required=True)
    key_prefix = fields.String()


class MessageSourceSchema(StrictSchema):
    """Configuration schema for a message source"""
    orgs = fields.Dict(keys=fields.String(),
                       values=fields.Nested(SalesforceOrgSchema()),
                       required=True,
                       validate=Length(min=1),
                       attribute="org_specs")
    replay = fields.Nested(ReplaySchema(), attribute="replay_spec")


class AmqpExchangeSchema(StrictSchema):
    """Configuration schema for declaring AMQP exchanges"""
    exchange_name = fields.String(required=True, validate=Length(min=1))
    type_name = fields.String(required=True, validate=OneOf(["fanout",
                                                             "direct",
                                                             "topic",
                                                             "headers"]))
    passive = fields.Boolean(default=False)
    durable = fields.Boolean(default=False)
    auto_delete = fields.Boolean(default=False)
    no_wait = fields.Boolean(default=False)
    arguments = fields.Dict(allow_none=True)


class AmqpBrokerSchema(StrictSchema):
    """Configuration schema for AMQP connection parameters"""
    host = fields.String(required=True, validate=Length(min=1))
    port = fields.Int(default=None, allow_none=True,
                      validate=Range(min=1, max=(2 ** 16) - 1))
    login = fields.String(default="guest")
    password = fields.String(default="guest")
    virtualhost = fields.String(default="/")
    ssl = fields.Boolean(default=False)
    verify_ssl = fields.Boolean(default=True)
    login_method = fields.String(default="AMQPLAIN")
    insist = fields.Boolean(default=False)
    exchanges = fields.List(fields.Nested(AmqpExchangeSchema()),
                            required=True,
                            validate=Length(min=1),
                            attribute="exchange_specs")


class MessageSinkSchema(StrictSchema):
    """Configuration schema for a message sink"""
    brokers = fields.Dict(keys=fields.String(),
                          values=fields.Nested(AmqpBrokerSchema()),
                          required=True,
                          validate=Length(min=1),
                          attribute="broker_specs")


class RouteSchema(StrictSchema):
    """Configuration schema for route parameters"""
    broker_name = fields.String(required=True, validate=Length(min=1))
    exchange_name = fields.String(required=True)
    routing_key = fields.String(required=True, validate=Length(min=1))
    properties = fields.Dict(
        keys=fields.String(
            validate=OneOf(("content_type",
                            "content_encoding",
                            "headers",
                            "delivery_mode",
                            "priority",
                            "correlation_id",
                            "reply_to",
                            "expiration",
                            "message_id",
                            "timestamp",
                            "type",
                            "user_id",
                            "app_id",
                            "cluster_id"))
        ),
        allow_none=True
    )


class RoutingRuleSchema(StrictSchema):
    """Configuration schema for routing rule"""
    condition = fields.String(required=True, validate=Length(min=1),
                              attribute="condition_spec")
    route = fields.Nested(RouteSchema(), required=True, attribute="route_spec")


class MessageRouterSchema(StrictSchema):
    """Configuration schema for message router"""
    default_route = fields.Nested(RouteSchema(),
                                  required=True,
                                  allow_none=True,
                                  attribute="default_route_spec")
    rules = fields.List(fields.Nested(RoutingRuleSchema()),
                        missing=list,
                        attribute="rule_specs")


class ApplicationConfigSchema(StrictSchema):
    """Congiguration schema for setting up the complete rabbit_force
    application"""
    source = fields.Nested(MessageSourceSchema(), required=True)
    sink = fields.Nested(MessageSinkSchema(), required=True)
    router = fields.Nested(MessageRouterSchema(), required=True)


def get_config_loader(file_path):
    """Find the appropriate config loader for *file_path*

    :param str file_path: An absolute or relative file path
    :return: A callable capable of loading the config file
    :rtype: :func:`callable` or None
    """
    # get the file's suffix without the period
    path = Path(file_path)
    suffix = path.suffix.lower()
    if suffix:
        suffix = suffix[1:]

    # return the json or the yaml load function
    if suffix == "json":
        LOGGER.debug("Using JSON config loader for %r", file_path)
        return json.load
    elif suffix in ("yml", "yaml"):
        LOGGER.debug("Using YAML config loader for %r", file_path)
        return yaml.safe_load

    # if the suffix of the file is not recognized return None
    LOGGER.debug("No config loader found for %r", file_path)
    return None


def load_config(file_path):
    """Load and validate the application's configuration from *file_path*

    :param str file_path: An absolute or relative file path
    :return: The loaded configuration
    :rtype: dict
    :raise ConfigurationError: If the file_path's suffix is not recognized, \
    if the file can't be loaded or if validation of the file's contents fail.
    """
    # get the appropriate loader for the file_path
    loader = get_config_loader(file_path)
    if not loader:
        raise ConfigurationError(f"Unrecognized configuration file "
                                 f"format for {file_path!r}.")

    # load the config file's contents
    LOGGER.debug("Loading configuration from %r", file_path)
    try:
        with open(file_path, "rt") as file:
            unvalidated_config = loader(file)
    except Exception as error:
        raise ConfigurationError(f"Failed to load configuration "
                                 f"file {file_path!r}. {error!s}") from error

    # validate the contents of the file
    LOGGER.debug("Validating configuration")
    try:
        config = ApplicationConfigSchema().load(unvalidated_config)
    except ValidationError as error:
        raise ConfigurationError(f"Failed to validate configuration "
                                 f"file {file_path!r}. {error!s}") from error

    return config
